### PR TITLE
[DATA-1892] Migrate the eight legacy-writer election-api tables to the swap DAG

### PR DESCRIPTION
**Pull request by danpelota**
## What

Moves the eight tables still delivered by the legacy JDBC upsert writer (`write__election_api_db.py`) onto the `sync_election_api` DAG, reworks delivery to a set-wise atomic swap, and disables the writer. All 13 election-api tables are staged from their dbt marts, then one transaction renames the entire set into place.

## How

- Each table is a declarative `MartSync` entry (spec, mart, gate, FK parents) consumed by a single task-group factory: build staging, load, add generated PK/index/FK DDL, quality gate. Every Prisma constraint name is declared exactly once.
- No column lists or row transforms: the load selects the live Postgres table's own columns (minus a small db-owned allowlist) from the mart, so dbt must publish matching names and types. A live column the mart lacks fails the load; a new live column whose data the mart already publishes starts flowing automatically. What the old writer transformed now lives in the marts: zip_to_position mints its deterministic id and updated_at, person and office_holder stamp build timestamps, position publishes br_database_id as string to match the Postgres text column (numeric consumers cast back explicitly).
- Staging FKs reference sibling staging tables, never live ones, so the staging set is self-contained and validates against its own vintage. The swap is one transaction: live set renamed aside, staging set renamed in, constraints riding along. No child rows are ever mutated and the API only ever sees one complete, referentially consistent vintage. Leftover `_old` tables from a crashed run are pre-dropped inside the swap transaction.
- Quality gates: count ratio vs prior live, cold-start floor, staged-vs-live id-overlap floor on the Prisma-graph tables, NULL probes, plus two per-table extras (ZipToPosition state coverage, Projected_Turnout natural-key uniqueness).
- The only cross-group ordering is `parent.build_indexes_and_fk >> child.build_indexes_and_fk`; loads run in parallel. The single rehearsal gate sits after all 13 quality checks, so a skip affects only swap/drop_old.
- `m_election_api__candidacy` now emits `person_id` directly, guarded against `m_election_api__person`.

## Cutover plan

1. Merge. The writer is disabled by this PR (`+enabled: false`; code removal is a follow-up), and the DAG deploys in rehearsal mode behind the `election_api_swap_enabled` Variable: all 13 staging tables built, loaded, indexed, and gated nightly; only the swap is withheld. Nothing delivers to election-api during rehearsal, so keep the window to a night or two.
2. One-time, right after merge: trigger `dbt run --select m_election_api__position --full-refresh` in prod. The mart is incremental and ignores the br_database_id bigint-to-string type change without it (the position load fails loudly until this runs).
3. Verify a rehearsal night in dev and prod: staging populated, gates green.
4. Flip `election_api_swap_enabled` to `true`. Step 1 must have actually deployed first — see the lock-contention finding below, the disabled writer is a prerequisite, not just a tidy-up.
5. Follow-up PR: delete `write__election_api_db.py` and its tests, retire the dedicated JDBC all-purpose cluster, drop `m_election_api__place.source_updated_at` (writer-only watermark), and clean up the stale `election_api_race_swap_enabled` Variable in Astro.

Notes for the first enabled run:
- The swap prunes rows the writer could never delete (Stance carries ~23k stale rows, District ~5k). Row counts drop accordingly; the gates allow this, and the id-overlap floor (0.90) was checked against current prod counts (worst case District at 0.96).
- The gp_candidate_id fill-rate test on the candidacy mart currently fails on a fresh build (5,817 nulls vs the 1,000 error threshold; prod's last build has 519). This predates this PR (upstream ER crosswalk drift) and will hit the next scheduled prod job independently.

## Dev end-to-end validation

Ran the whole lifecycle against `astro-dev` and the dev election-api Postgres with the swap **enabled**, sourcing marts from a dev schema via `election_api_source_schema`. All 55 tasks green; swap and drop_old both completed.

Post-swap Postgres row counts match their source marts exactly, all 13 tables: DistrictTopIssue 5,098,997 / ZipToPosition 1,356,467 / Race 1,013,467 / Projected_Turnout 1,004,060 / Position 286,419 / Person 261,777 / Candidacy 246,722 / Stance 215,739 / OfficeHolder 139,634 / District 131,340 / Place 48,603 / Elected_Office_Support 599 / Issue 20.

Structure verified directly in Postgres: 13 canonical PKs (`Elected_Office_Support` correctly on `elected_office_id`), all 14 FKs `convalidated` and pointing at live public siblings, per-table index counts matching every declared spec, zero surviving `_new`/`_old` names, staging schema empty. `Position.br_database_id` is `text`, Person/OfficeHolder build timestamps non-null, ZipToPosition ids non-null across 51 states, DistrictTopIssue issue-level flags non-null. `Candidacy.person_id` populated on 136,658 of 246,722 rows, which is the guarded join behaving as documented. All 13 quality gates passed at their declared floors, including the 0.90 id-overlap.

Two operational findings worth reviewing:

**The legacy writer blocks the swap.** Its `DELETE FROM "Candidacy" WHERE race_id not in (select id from "Race")` (`write__election_api_db.py:753`) held ROW EXCLUSIVE on Candidacy for over an hour on dev — five instances stacked up — while the swap's `LOCK TABLE ... ACCESS EXCLUSIVE` gives up after 10s. Four consecutive swap attempts failed on lock timeout until those sessions were terminated. This makes cutover ordering a hard dependency: the writer has to be genuinely off before `election_api_swap_enabled` flips, otherwise the swap fails every night on lock timeout while everything upstream of it stays green. The 10s timeout itself is right — an `ACCESS EXCLUSIVE` request queues ahead of new readers, so waiting longer would stall the API rather than wait politely.

**Parallel loads can OOM the worker.** `race` and `candidacy` load_staging were both SIGKILLed on their first attempt with four large loads running concurrently; both passed on retry. Per-task memory is bounded to one partition, so this is the sum across parallel tasks, not a per-task regression. With retries now at 1 that leaves exactly one retry of headroom. If it recurs, the fix is an Airflow pool capping concurrent loads rather than restoring retries.

## Also in this revision

- Task retries 3 → 1. A step that fails twice is not a transient blip, and a fast red run surfaces it instead of burying it in a retry chain.
- Comment pass over the DAG and utils: removed duplication between the utils module docstring and the DAG's rendered `doc_md`, and inline comments that restated the docstring a few lines above (net -63 lines). Kept the non-obvious parts — Prisma naming convention, why the `_old` pre-drop exists, why the regclass identifiers need quoting.

## Testing

- Airflow: 79 tests pass, including a FakePostgres crash-point sweep over the multi-table swap transaction (with leftover `_old` sets holding FKs to each other), committed-swap-retry fail-closed, and self-referencing FK ride-through. DAG parses under real Airflow: 55 tasks, no import errors, `dag.validate()` clean.
- dbt: all changed models built and tested in dev (90 passes; the one error is the pre-existing gp_candidate_id drift). Position full-refreshed in dev to apply the type change; every mart verified to cover its live Postgres column set exactly.

### CODE REVIEW

**Comment by danpelota** [comment]
delegate review

**Comment by danpelota** [comment]
Hold review: reworking to a single set-wise atomic swap of all 13 tables per feedback (removes the inbound-FK re-point and orphan DELETE machinery entirely). Will re-request when pushed.

**Comment by danpelota** [comment]
delegate review

**Comment by danpelota** [comment]
delegate review

**Comment by danpelota** [comment]
delegate review

**Comment by danpelota** [comment]
delegate review

File: airflow/astro/dags/sync_election_api.py [hunk 1]
```diff
@@ -115,1024 +116,445 @@ def _open_pg():
...
+        partition_column="issue",
+        gate=QualityGate(
+            cold_start_floor=100_000,
+            db_owned_columns=frozenset({"created_at"}),
+            # The mart's LEFT JOIN to haystaq_issue_tags only emits NULL flags
+            # on drift; belt-and-suspenders over the dbt-side tests.
+            not_null_columns=("is_local", "is_regional", "is_state", "is_federal"),
+        ),
+        parents=("district",),
+    ),
```
[the remark this answers is not in the document]
Comment (danpelota) [reply airflow/astro/dags/sync_election_api.py:416]: Declining this one.

The check did exist in the old task, and the port dropped it deliberately-in-effect rather than by oversight — but it should stay dropped. It encoded an assumption we do not hold: that the issue taxonomy only ever grows. Upstream modeling drops whole issue categories regularly, and that is legitimate work, not a regression.

This DAG's contract is to reflect what the mart publishes. A gate that refuses the swap whenever the distinct-category count decreases would block normal upstream taxonomy changes and would need a hand override every time one lands, which makes it worse than no gate — it trains the operator to bypass the gates.

What still covers the failure modes that matter here: the row-count ratio (0.5) catches a collapse, and the `not_null_columns` probes on `is_local`/`is_regional`/`is_state`/`is_federal` catch the LEFT JOIN drift against `haystaq_issue_tags` that those flags exist to detect. Correctness of the taxonomy itself belongs in the dbt tests on the mart, where the taxonomy is actually defined.

File: dbt/project/dbt_project.yml [hunk 2]
```diff
@@ -90,6 +90,10 @@ models:
       write__people_api_db:
         # Disabled due to performance issues (timing out)
         +enabled: false
+      write__election_api_db:
+        # Superseded by the sync_election_api Airflow swap DAG; code removal
+        # is a follow-up once the cutover Variable is flipped.
+        +enabled: false
```
[the remark this answers is not in the document]
Comment (danpelota) [reply dbt/project/dbt_project.yml:96]: Option (a) is contraindicated; (b) is already in place.

**(a) leaving the writer enabled so it lands atomically with the Variable flip would break the swap.** From yesterday's dev cutover: the writer's `DELETE FROM public."Candidacy" WHERE race_id not in (select id from public."Race") or race_id is null` (`write__election_api_db.py:753`) holds ROW EXCLUSIVE on Candidacy for well over an hour — five instances were stacked up concurrently on dev, the oldest running 1h47m. The swap takes ACCESS EXCLUSIVE with `lock_timeout = 10s`, so four consecutive swap attempts failed on `LockNotAvailable` until those writer sessions were terminated, with all 13 staging tables built and every quality gate green.

So the writer being disabled is a **prerequisite** for the swap succeeding at all, not a tidy-up that can land alongside it. In the order (a) proposes, the swap would fail on lock timeout every night while everything upstream of it stayed green. The 10s timeout is itself correct — an ACCESS EXCLUSIVE request queues ahead of new readers, so waiting longer would stall the API rather than wait politely.

**(b) is what we are doing.** The cutover plan in the PR description gives the ordered steps, and step 4 now states this dependency explicitly.

On the freeze being unenforced: the rehearsal default of `"false"` is deliberate. A merge must not silently begin swapping 13 tables wholesale — that is the property the default is protecting, and defaulting it on to avoid a freeze would be the worse failure. Bounding the window is a process guarantee, which is where it belongs; the DAG cannot know when the operator intends to cut over. The full lifecycle including the swap has now been validated end to end in dev, so the rehearsal window exists to confirm prod parity, not to discover unknowns.

**Comment by danpelota** [comment]
delegate review

**Comment by danpelota** [comment]
delegate review

File: airflow/astro/dags/sync_election_api.py [hunk 3]
```diff
@@ -1,35 +1,49 @@
...
-5. **swap** — atomic rename swap (`_new` -> live, live -> `_old`) with all
-   indexes and constraints renamed to canonical Prisma names. A leftover
-   `_old` from a run that crashed before drop_old is pre-dropped in the same
-   transaction, so a crashed run never wedges subsequent ones.
-6. **drop_old** — drop the renamed-aside `_old` table.
-
-The race group adds a seventh step: a cutover short-circuit between
-quality_checks and swap (rehearsal mode until the cutover Variable flips).
+   The column list is the live table's own columns (minus any
+   `db_owned_columns`), so the mart must publish matching names and types —
```
Comment (danpelota) [inline airflow/astro/dags/sync_election_api.py:12]: I'm going to remove the `db_owned_columns` in a followup PR as well. I think the expectation is this loader DAG should be responsible for loading a 1:1 copy of each table from Databricks into Postgres.

File: airflow/astro/dags/sync_election_api.py [hunk 4]
```diff
@@ -115,1033 +116,445 @@ def _open_pg():
...
+            f") AS dupe_keys"
+        )
+        dup_keys = cur.fetchone()[0]
+    finally:
+        cur.close()
+    if dup_keys > 0:
+        raise ValueError(
+            f"{dup_keys} duplicate (district_id, election_year, "
+            f"election_code) keys in staging — refusing to swap"
+        )
```
[the remark this answers is not in the document]
Comment (danpelota) [reply airflow/astro/dags/sync_election_api.py:176]: Covered, but implemented differently than proposed — the proposal would not catch the risk it names.

Extracting a pure `check_no_duplicate_keys(dup_keys, ...)` can only assert `dup_keys > 0`. The wrong-grouping-column scenario you raise lives in the SQL, which stays in the impure wrapper either way, so a predicate over an already-computed count cannot see it. And unlike the id-overlap gate — which has a real boundary worth pinning (ratio vs floor, exactly-at-floor, and the no-declared-floor skip) — "any duplicate refuses" has no edge case, so that test would restate the implementation.

Instead `fa815c9d` drives `_pt_extra_checks` and `_ztp_extra_checks` against a fake cursor, asserting both directions. That exercises the real function body including the query and the `fetchone()[0]` read, so it covers the wiring as well as the decision. `_ztp_extra_checks` never had a test either; it does now.

Two notes on your surrounding observations:

You are right that there is no UNIQUE constraint on (district_id, election_year, election_code) in the spec, and that is the sharper point — a unique index would enforce this at the database level and fail at `build_indexes_and_fk`, which is strictly stronger than a pre-swap count. Not doing it here: Projected_Turnout rows re-key on model-version supersessions, so I want to confirm the mart can never legitimately emit two model_versions for one key before converting a clear gate failure into an index-build failure. Worth a follow-up.

Separately, the old `_pt_quality_gate` took its ratio baseline from the prior live table's DISTINCT natural-key count, whereas the generic `check_counts` uses raw row count. If prod's live Projected_Turnout carries duplicate-inflated rows from the legacy writer, the raw-row baseline is larger and the ratio correspondingly smaller. That fails closed — it would refuse a swap rather than allow a bad one — so it is not a safety regression, but it is a first-enabled-run consideration.

**Comment by danpelota** [comment]
delegate review
